genie-web/src/main/java/com/netflix/genie/web/data/entities/listeners/JobEntityListener.java

```diff
@@ -0,0 +1,101 @@
+/*
+ *
+ *  Copyright 2019 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.data.entities.listeners;
+
+import com.netflix.genie.common.dto.JobStatus;
+import com.netflix.genie.web.data.entities.JobEntity;
+import com.netflix.genie.web.data.observers.PersistedJobStatusObserver;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nullable;
+import javax.persistence.PostLoad;
+import javax.persistence.PostUpdate;
+
+/**
+ * Listener for Job JPA entity ({@link JobEntity}).
+ * Currently tracks persistent changes to the status of a job and notifies an observer.
+ * Could be extended to do proxy more persist changes.
+ * <p>
+ * N.B. Spring configuration.
+ * - This class does not appear in any AutoConfiguration as bean.
+ * It is referenced as {@link java.util.EventListener} by {@link JobEntity}.
+ * EntityManager creates it even if a bean of the same type exists already.
+ * - The constructor parameter {@code persistedJobStatusObserver} is marked {@code Nullable} so that an instance can be
+ * created even if no bean of that type is configured.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class JobEntityListener {
+
+    private final PersistedJobStatusObserver persistedJobStatusObserver;
+
+    /**
+     * Constructor (autowired during Persistence / EntityManager initialization).
+     *
+     * @param persistedJobStatusObserver the observer to notify
+     */
+    public JobEntityListener(
+        @Nullable final PersistedJobStatusObserver persistedJobStatusObserver
```

Comment:
Any reason you don't just take in a `List` or a `Set`?

Reply:
The idea is that this one takes exactly one observer (tho we may later add other kinds of observers and notifies it.
The observer is the thing that may fanout to different other components.
This double decoupling is to make sure we have a place where it's easy to inject (observers) and a place that is low-level (JPA entity listener)